### ISSUE TITLE
Testing Typscript

### ISSUE THREAD

**Issue by hipstersmoothie** [open-wc/open-wc#485]
Hey, I love how easy this was all to set up. Converting to typescript was even a breeze for the most part. 

However, I can't seem to get karma working with my typescript. I've tried configuring `karma-typescript` myself but have failed for the last few hours. 

I would really like to see some docs for setting up tests for typescript. The docs are great in most other places, but lack here.

**Comment by hipstersmoothie** [comment]
Findings:

If i use karma-typescript with the following config things kinda start to work

```js
const createDefaultConfig = require('@open-wc/testing-karma/default-config');
const merge = require('webpack-merge');

module.exports = config => {
  config.set(
    merge(createDefaultConfig(config), {
      frameworks: ['karma-typescript'],
      reporters: ['karma-typescript'],
      preprocessors: {
        '**/*.ts': ['esm', 'karma-typescript']
      },
      files: [
        {
          pattern: config.grep ? config.grep : 'src/**/*.ts',
          type: 'module'
        }
      ]
    })
  );

  return config;
};
```

This will error on line 3 of the following test file

```ts
import { html, fixture, expect } from '@open-wc/testing';

import './visually-hidden';

describe('<visually-hidden>', () => {
  it('allows property heading to be overwritten', async () => {
    const el = await fixture(html`
      <visually-hidden></visually-hidden>
    `);

    expect(el.heading).to.equal('different heading');
  });
});
```

I can fix this be adding `.ts` to the import (that breaks my editor plugin though lit-plugin). But it still doesn't build. I am presented with the following error that I don't know how to fix.

```sh
08 06 2019 11:44:14.252:INFO [compiler.karma-typescript]: Compiling project using Typescript 3.5.1
08 06 2019 11:44:18.114:INFO [compiler.karma-typescript]: Compiled 4 files in 3849 ms.
08 06 2019 11:44:18.700:ERROR [karma-server]: Error: Invalid syntax in bundle: 'import' and 'export' may only appear at the top level (23:0)
in /var/folders/kd/4dz51_q566sg7y02l6gwc8zr9mjqwm/T/karma-typescript-bundle-30316F5TV217rQ9xW.js
at line 23, column 0:

... import { TemplateResult } from 'lit-html';
import  ...

Possible fix: configure karma-typescript to compile es6 modules with karma-typescript-es6-transform

    at Validator.validate (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/dist/bundler/validator.js:20:23)
    at Bundler.writeMainBundleFile (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/dist/bundler/bundler.js:183:24)
    at /Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/dist/bundler/bundler.js:118:19
    at wrapper (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:270:20)
    at replenish (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:437:29)
    at iterateeCallback (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:426:21)
    at /Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:323:20
    at /Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/dist/bundler/globals.js:25:17
    at wrapper (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:270:20)
    at replenish (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:437:29)
    at iterateeCallback (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:426:21)
    at /Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:323:20
    at /Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/dist/bundler/globals.js:20:21
    at onDependenciesResolved (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/dist/bundler/resolve/resolver.js:71:20)
    at wrapper (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:270:20)
    at iteratorCallback (/Users/alisowski/Documents/visually-hidden/node_modules/karma-typescript/node_modules/async/dist/async.js:497:17
```

**Comment by hipstersmoothie** [comment]
I feel like the esm loader should be handling this file but maybe it ignores node_modules?

**Comment by hipstersmoothie** [comment]
Adding 

```
karmaTypescriptConfig: {
    bundlerOptions: {
        validateSyntax: false
    }
}
```

get me further but i think it still breaks on the import line

```sh
08 06 2019 11:51:27.321:INFO [compiler.karma-typescript]: Compiling project using Typescript 3.5.1
08 06 2019 11:51:31.722:INFO [compiler.karma-typescript]: Compiled 4 files in 4385 ms.
08 06 2019 11:51:32.305:INFO [bundler.karma-typescript]: Bundled imports for 3 file(s) in 80 ms.
08 06 2019 11:51:32.395:INFO [karma-server]: Karma v4.1.0 server started at http://0.0.0.0:9881/
08 06 2019 11:51:32.395:INFO [launcher]: Launching browsers ChromeHeadlessNoSandbox with concurrency unlimited
08 06 2019 11:51:32.402:INFO [launcher]: Starting browser ChromeHeadless
08 06 2019 11:51:32.899:INFO [HeadlessChrome 74.0.3729 (Mac OS X 10.14.0)]: Connected on socket peH-f4pwaZmdAO0TAAAA with id 91106665
HeadlessChrome 74.0.3729 (Mac OS X 10.14.0) ERROR
  Uncaught SyntaxError: Unexpected token {
  at /var/folders/kd/4dz51_q566sg7y02l6gwc8zr9mjqwm/T/karma-typescript-bundle-33157M2yyXTngV4Jk.js:23:8
  
  SyntaxError: Unexpected token {
```

**Comment by hipstersmoothie** [comment]
Currently I have tests working using most of the method described in https://github.com/open-wc/open-wc/issues/252 (tests in JS, importing built TS). But this is less than ideal. I really want tests in TS importing TS source.

Here is the project I'm working on: https://github.com/open-wc/open-wc/issues/252

**Comment by Ben-Barron** [comment]
+1

I've spent the last few days playing about - but I can't get a working configuration :(

**Comment by thepassle** [comment]
I poked around with this a bit yesterday, and I had some success with this karma config: https://gist.github.com/thepassle/fd1f30ddfd33de105df90d96381374b1

Could you let me know if that works for you?

**Comment by LarsDenBakker** [comment]
A regular preprocessor will not work because they are only run for the initial entry modules, subsequent module imports made by the browser won't be run through the preprocessor because they are served by the `karma-esm` middleware.

We have to think about how we want to support this. Some ideas:
1. Support reading the user's `.babelrc`, and advise using typescript babel plugin.
  - This would limit us to only supporting babel plugins, but it would make it straightforward and simple
2. Run all served modules through the preprocessor
  - I am not sure if this is technically feasible
3. Advice using `karma-webpack` for any non-standard code

**Comment by hipstersmoothie** [comment]
> Support reading the user's .babelrc, and advise using typescript babel plugin.

You could even have this on the default. I don't think it would change much

**Comment by daKmoR** [comment]
> Advice using karma-webpack for any non-standard code

sounds like a good plan to me 👍 

if you need to build you should build - and we have webpack for that (for testing that should be good enough I would say)

**Comment by LarsDenBakker** [comment]
According to that logic we should pick option 1. Webpack will do bundling in addition to building, babel only does building, right?

**Comment by hipstersmoothie** [comment]
My use case isn't at all complicated at this point. I literally just changes .js to .ts and now it's hard to test, so webpack seems like quite a large hammer for this problem

**Comment by LarsDenBakker** [comment]
Well the change is not so small since you are going from something that runs natively in the browser to something that doesnt. 

Javascript build systems get messy pretty quickly so we do have to be cafeful.

**Comment by mkay581** [comment]
@hipstersmoothie were you ever able to get this working with Typescript files? I'm running into your exact same issues above.

**Comment by hipstersmoothie** [comment]
nope was only able to run the dist files easily

**Comment by LarsDenBakker** [comment]
Update on this topic: we're working on a new dev server (https://github.com/open-wc/open-wc/pull/520) which we're planning to use to power our karma config as well.

When this is done, we can use regular karma preprocessors again so it will be much easier to use typescript.

**Comment by mkay581** [comment]
Hey @LarsDenBakker I may be missing something, but how does the new dev server help with this issue?

**Comment by Nachasic** [comment]
I've also been struggling trying to set up typescript tests. Encountered every problem mentioned in this thread myself.

I also feel strongly opposed to the idea of using `karma-webpack`, because in general case it pretty much marries you to webpack, forcing you to maintain multiple bundling configurations, especially if you're already using something else, i.e. rollup or (in my case) fuse-box. IMHO webpack is a slow and sluggish pile of tech debt, so I would like to stay as far away from it as possible.

If bundling tests for Karma is the only way to go (looks like it), I would appreciate well-documented recommendations on how to do it instead, so I can implement it myself. Subscribing to this thread in hopes of helping out with anything.

**Comment by mkay581** [comment]
Yeah, outside of compiling TS, we shouldn't need webpack (or any bundler really) if we're just running the compiled javascript with features that are already natively supported in a modern browser like Chrome. 

That original `lit-html` error that @hipstersmoothie mentioned above shouldn't be an error when running it in Chrome using karma because I _believe_ I was told in the lit slack channel (by @LarsDenBakker I think) that lit doesn't need to be bundled to run in modern browsers.

**Comment by LarsDenBakker** [comment]
I share your feeling of not wanting to use webpack or any bundler during development/testing. That's a primary goal for us as well. The core problem is that for many tools running code unbundled does not work out of the box. So while things work just fine when you run it in the browser, it doesn't when you use some tools.

The problem with karma is that it doesn't serve any files it didn't know about beforehand. Because we want to use the module loader of the browser, we can't and shouldn't know the dependency tree beforehand. 

It took quite some effort getting this to work, but we did it with `karma-esm`. Unfortunately this has the side effect of not running the karma preprocessors since we're bypassing karma for serving files. 

So to summarize:
- There is a solution for running karma with native modules (our `karma-esm` plugin)
- There is a solution for running karma with typescript (karma typescript preprocessor, karma webpack etc.)

But they don't both work together yet.

**Comment by mkay581** [comment]
Ah I see. So should we open a new issue specifically to have karma-esm work with karma's preprocessors?

**Comment by LarsDenBakker** [comment]
That's something to investigate, either we will do that or special-case typescript support if it's going to be too much effort.

**Comment by LarsDenBakker** [comment]
We've documented a testing approach with the new version of our testing setup. I'm curious to hear if this works for you guys.

Unfortunately a solution which utilizes the typescript compiler is not possible because our solution is based on serving modules one by one, while the typescript compiler requires an analysis of your entire project.

**Comment by heruan** [comment]
Let's try to sum this up: as of the latest PRs, which steps should I take to test my TypeScript sources?

**Comment by LarsDenBakker** [comment]
We have a section in the docs for this now

**Comment by thepassle** [comment]
https://open-wc.org/testing/karma-esm.html#testing-typescript
